File: pkg/client/stream.go
package client

import (
	"context"
	"fmt"
	"sync"
)

// StreamWriter provides a streaming multipart upload API where individual
// parts can be submitted concurrently as they become available.
// It wraps the v2 upload protocol (initiate → presign → upload → complete),
// with fallback to v1 if the server doesn't support v2.
//
// Usage:
//  1. Call Client.NewStreamWriter() to create a StreamWriter (lazy — no network call yet).
//  2. Call WritePart() for each part as data becomes ready (concurrency-safe).
//  3. Call Complete() to upload the final part and finalize the upload.
//  4. Call Abort() on any error to clean up server-side state.
type StreamWriter struct {
	client   *Client
	path     string
	totalSize int64

	mu        sync.Mutex
	plan      *uploadPlanV2    // lazily initialized on first WritePart
	uploaded  map[int]completePart // partNumber → completed part
	inflight  sync.WaitGroup
	err       error            // first error from any goroutine
	started   bool
	completed bool

	sem chan struct{} // concurrency limiter
}

// NewStreamWriter creates a StreamWriter for streaming multipart upload.
// No network call is made until the first WritePart.
func (c *Client) NewStreamWriter(ctx context.Context, path string, totalSize int64) *StreamWriter {
	return &StreamWriter{
		client:    c,
		path:      path,
		totalSize: totalSize,
		uploaded:  make(map[int]completePart),
		sem:       make(chan struct{}, uploadMaxConcurrency),
	}
}

// Started reports whether the upload has been initiated.
func (sw *StreamWriter) Started() bool {
	sw.mu.Lock()
	defer sw.mu.Unlock()
	return sw.started
}

// initLocked initiates the multipart upload. Caller must hold sw.mu.
func (sw *StreamWriter) initLocked(ctx context.Context) error {
	if sw.started {
		return nil
	}
	plan, err := sw.client.initiateUploadV2(ctx, sw.path, sw.totalSize)
	if err == errV2NotAvailable {
		return fmt.Errorf("streaming upload requires v2 protocol: %w", err)
	}
	if err != nil {
		return fmt.Errorf("initiate stream upload: %w", err)
	}
	sw.plan = plan
	sw.started = true
	return nil
}

// WritePart uploads a single part in the background. partNum is 1-based.
// data is copied internally so the caller may reuse the buffer after return.
// This method is concurrency-safe.
func (sw *StreamWriter) WritePart(ctx context.Context, partNum int, data []byte) error {
	sw.mu.Lock()
	if sw.err != nil {
		err := sw.err
		sw.mu.Unlock()
		return err
	}
	if sw.completed {
		sw.mu.Unlock()
		return fmt.Errorf("stream writer already completed")
	}

	// Lazy init
	if err := sw.initLocked(ctx); err != nil {
		sw.err = err
		sw.mu.Unlock()
		return err
	}
	plan := sw.plan
	sw.mu.Unlock()

	// Copy data so caller can reuse buffer
	buf := make([]byte, len(data))
	copy(buf, data)

	// Acquire semaphore
	select {
	case sw.sem <- struct{}{}:
	case <-ctx.Done():
		return ctx.Err()
	}

	sw.inflight.Add(1)
	go func() {
		defer sw.inflight.Done()
		defer func() { <-sw.sem }()

		// Presign this part
		pp, err := sw.client.presignOnePart(ctx, plan.UploadID, partNum)
		if err != nil {
			sw.setError(fmt.Errorf("presign part %d: %w", partNum, err))
			return
		}

		// Upload
		etag, err := sw.client.uploadOnePartV2(ctx, *pp, buf)
		if err == errPresignExpired {
			// Retry with fresh presign
			pp2, err2 := sw.client.presignOnePart(ctx, plan.UploadID, partNum)
			if err2 != nil {
				sw.setError(fmt.Errorf("re-presign part %d: %w", partNum, err2))
				return
			}
			etag, err = sw.client.uploadOnePartV2(ctx, *pp2, buf)
		}
		if err != nil {
			sw.setError(fmt.Errorf("upload part %d: %w", partNum, err))
			return
		}

		sw.mu.Lock()
		sw.uploaded[partNum] = completePart{Number: partNum, ETag: etag}
		sw.mu.Unlock()
	}()

	return nil
}

// Complete waits for all inflight parts, uploads the final part if provided,
// and calls CompleteMultipartUpload.
// finalPartNum/finalPartData can be used for the last (possibly partial) part.
// If finalPartData is nil, no additional part is uploaded.
func (sw *StreamWriter) Complete(ctx context.Context, finalPartNum int, finalPartData []byte) error {
	// Upload final part synchronously if provided
	if finalPartData != nil && len(finalPartData) > 0 {
		sw.mu.Lock()
		if sw.err != nil {
			err := sw.err
			sw.mu.Unlock()
			return err
		}
		if err := sw.initLocked(ctx); err != nil {
			sw.err = err
			sw.mu.Unlock()
			return err
		}
		plan := sw.plan
		sw.mu.Unlock()

		pp, err := sw.client.presignOnePart(ctx, plan.UploadID, finalPartNum)
		if err != nil {
			return fmt.Errorf("presign final part %d: %w", finalPartNum, err)
		}
		etag, err := sw.client.uploadOnePartV2(ctx, *pp, finalPartData)
		if err != nil {
			return fmt.Errorf("upload final part %d: %w", finalPartNum, err)
		}
		sw.mu.Lock()
		sw.uploaded[finalPartNum] = completePart{Number: finalPartNum, ETag: etag}
		sw.mu.Unlock()
	}

	// Wait for all background parts
	sw.inflight.Wait()

	sw.mu.Lock()
	defer sw.mu.Unlock()

	if sw.err != nil {
		return sw.err
	}

	if !sw.started || sw.plan == nil {
		return fmt.Errorf("stream writer was never started")
	}

	// Build ordered parts list from actually-uploaded parts.
	// We use sw.uploaded (not sw.plan.TotalParts) because the file may have
	// grown after initiation, making TotalParts stale.
	if len(sw.uploaded) == 0 {
		return fmt.Errorf("no parts uploaded in stream upload")
	}
	// Find the max part number to determine the range.
	maxPart := 0
	for pn := range sw.uploaded {
		if pn > maxPart {
			maxPart = pn
		}
	}
	parts := make([]completePart, 0, maxPart)
	for i := 1; i <= maxPart; i++ {
		cp, ok := sw.uploaded[i]
		if !ok {
			return fmt.Errorf("missing part %d in stream upload (have %d parts, max %d)", i, len(sw.uploaded), maxPart)
		}
		parts = append(parts, cp)
	}

	sw.completed = true
	return sw.client.completeUploadV2(ctx, sw.plan.UploadID, parts)
}

// Abort cancels the multipart upload and cleans up server-side state.
func (sw *StreamWriter) Abort(ctx context.Context) error {
	sw.inflight.Wait()

	sw.mu.Lock()
	defer sw.mu.Unlock()

	if !sw.started || sw.plan == nil {
		return nil
	}
	return sw.client.abortUploadV2(ctx, sw.plan.UploadID)
}

// setError records the first error encountered.
func (sw *StreamWriter) setError(err error) {
	sw.mu.Lock()
	defer sw.mu.Unlock()
	if sw.err == nil {
		sw.err = err
	}
}

File: clients/drive9-py/drive9/stream.py
"""Streaming multipart upload helpers."""

from concurrent.futures import ThreadPoolExecutor
import threading

import requests

from .exceptions import Drive9Error
from .transfer import _V2NotAvailable

_UPLOAD_MAX_CONCURRENCY = 16


class StreamWriter:
    """Streaming multipart upload API where individual parts can be submitted
    concurrently as they become available.

    Wraps the v2 upload protocol (initiate → presign → upload → complete).
    """

    def __init__(self, client, path: str, total_size: int, expected_revision: int = -1):
        self._client = client
        self._path = path
        self._total_size = total_size
        self._expected_revision = expected_revision

        self._mu = threading.Lock()
        self._plan = None
        self._uploaded = {}
        self._inflight = 0
        self._cond = threading.Condition(self._mu)
        self._err = None
        self._started = False
        self._completed = False
        self._aborted = False
        self._closing = False

        self._executor = None

    @property
    def started(self) -> bool:
        with self._mu:
            return self._started

    def _init_locked(self):
        if self._started:
            return
        try:
            plan = self._client._initiate_upload_v2(
                self._path, self._total_size, self._expected_revision
            )
        except Exception as exc:
            msg = f"initiate stream upload: {exc}"
            if isinstance(exc, Drive9Error) and "v2 protocol" in str(exc):
                raise
            if isinstance(exc, _V2NotAvailable):
                raise Drive9Error("streaming upload requires v2 protocol: v2 upload API not available") from exc
            raise Drive9Error(msg) from exc
        self._plan = plan
        self._started = True

    def write_part(self, part_num: int, data: bytes) -> None:
        """Upload a single part in the background. part_num is 1-based.

        Data is copied internally so the caller may reuse the buffer after return.
        """
        if part_num < 1:
            raise Drive9Error("part number must be >= 1")

        with self._mu:
            if self._err is not None:
                raise self._err
            if self._completed:
                raise Drive9Error("stream writer already completed")
            if self._aborted:
                raise Drive9Error("stream writer already aborted")
            if self._closing:
                raise Drive9Error("stream writer is closing")
            if part_num in self._uploaded:
                raise Drive9Error(f"part {part_num} already uploaded")

            self._init_locked()
            plan = self._plan

            total_parts = plan.get("total_parts")
            if total_parts is not None and part_num > total_parts:
                raise Drive9Error(
                    f"part number {part_num} exceeds total_parts {total_parts}"
                )

            self._inflight += 1

            if self._executor is None:
                self._executor = ThreadPoolExecutor(max_workers=_UPLOAD_MAX_CONCURRENCY)

        buf = bytes(data)

        def do_upload():
            try:
                pp = self._client._presign_one_part(plan["upload_id"], part_num)
                etag = self._client._upload_one_part_v2(plan["upload_id"], pp, buf)
                with self._mu:
                    self._uploaded[part_num] = {"number": part_num, "etag": etag}
            except Exception as exc:
                self._set_error(Drive9Error(f"upload part {part_num}: {exc}"))
            finally:
                with self._mu:
                    self._inflight -= 1
                    self._cond.notify_all()

        self._executor.submit(do_upload)

    def complete(self, final_part_num: int = 0, final_part_data: bytes = b"") -> None:
        """Wait for inflight parts, upload the final part (if provided), and finalize."""
        with self._mu:
            if self._completed:
                raise Drive9Error("stream writer already completed")
            if self._aborted:
                raise Drive9Error("stream writer already aborted")
            if self._closing:
                raise Drive9Error("stream writer is closing")
            self._closing = True

        self._wait_inflight()

        with self._mu:
            if self._err is not None:
                raise self._err
            if not self._started or self._plan is None:
                raise Drive9Error("stream writer was never started")
            plan = self._plan

        if final_part_data:
            pp = self._client._presign_one_part(plan["upload_id"], final_part_num)
            etag = self._client._upload_one_part_v2(plan["upload_id"], pp, final_part_data)
            with self._mu:
                self._uploaded[final_part_num] = {
                    "number": final_part_num,
                    "etag": etag,
                }

        with self._mu:
            if not self._uploaded:
                raise Drive9Error("no parts uploaded in stream upload")

            max_part = max(self._uploaded.keys())
            parts = []
            for i in range(1, max_part + 1):
                cp = self._uploaded.get(i)
                if cp is None:
                    raise Drive9Error(
                        f"missing part {i} in stream upload "
                        f"(have {len(self._uploaded)} parts, max {max_part})"
                    )
                parts.append(cp)

            self._completed = True
            upload_id = plan["upload_id"]

        try:
            self._client._complete_upload_v2(upload_id, parts)
        finally:
            self._shutdown_executor()

    def abort(self) -> None:
        """Cancel the multipart upload and clean up server-side state."""
        with self._mu:
            if self._aborted:
                return
            self._closing = True

        self._wait_inflight()

        with self._mu:
            self._aborted = True
            if not self._started or self._plan is None:
                self._shutdown_executor()
                return
            upload_id = self._plan["upload_id"]
        try:
            self._client._abort_upload_v2(upload_id)
        finally:
            self._shutdown_executor()

    def _wait_inflight(self):
        with self._mu:
            while self._inflight > 0:
                self._cond.wait(timeout=1.0)

    def _shutdown_executor(self):
        with self._mu:
            if self._executor is not None:
                self._executor.shutdown(wait=True)
                self._executor = None

    def _set_error(self, err):
        with self._mu:
            if self._err is None:
                self._err = err
